### PR TITLE
Add `undo` and `redo` command functionality

### LINKED ISSUE #76: Add `undo` command
As a user I want to undo my command so that I can easily amend commands inputted by mistake.

### LINKED ISSUE #84: Add `redo` command
As a user I want to redo my command so that I can amend wrong undo commands.

### PR DESCRIPTION
**Pull request by CJerrong**
Main changes in this pr:
- Support new undo and redo commands.
- Added a new datastoreVersionStorage which is in charge of storing snapshots of the datastore whenever it is mutated.
- Refactor commands that mutate the datastore to commit the datastore to the datastoreVersionStorage after completion.
- Update the developer guide regarding implementation of undo and redo functionality

### CODE REVIEW

File: src/main/java/scrolls/elder/model/DatastoreVersionStorage.java [hunk 1]
```diff
@@ -0,0 +1,56 @@
...
+        if (pointer < size) {
+            return true;
+        }
+
+        return false;
+
+    }
+
+    public ReadOnlyDatastore executeUndo() {
+        assert this.canUndo() : "Undo command cannot be executed";
```
Comment (chaaaaun) [inline src/main/java/scrolls/elder/model/DatastoreVersionStorage.java:70]: Actually this prompted me to do a little searching. According to [this post](https://stackoverflow.com/questions/2758224/what-are-assertions-in-java-and-when-should-they-be-used), assertions aren't even turned on by default. Should we operate on the assumption that they don't trigger?

File: src/main/java/scrolls/elder/model/DatastoreVersionStorage.java [hunk 2]
```diff
@@ -50,7 +50,16 @@ public ReadOnlyDatastore executeRedo() {
     }
 
     public void commitDatastore(ReadOnlyDatastore datastore) {
-        this.datastoreVersions.add(datastore);
+        int size = this.datastoreVersions.size();
+
+        // If not at end of list, purge the data before adding new datastore snapshot
```
Comment (chaaaaun) [inline src/main/java/scrolls/elder/model/DatastoreVersionStorage.java:100]: Good catch, we might need a couple of tests for various sequences of undo and redo to catch any edge cases.

File: src/main/java/scrolls/elder/logic/commands/PairCommand.java [hunk 3]
```diff
@@ -118,6 +118,7 @@ public CommandResult execute(Model model) throws CommandException {
 
         store.setPerson(personToPair1, newPerson1);
         store.setPerson(personToPair2, newPerson2);
+        model.commitDatastore();
         store.updateFilteredPersonList(Model.PREDICATE_SHOW_ALL);
```
Comment (chaaaaun) [inline src/main/java/scrolls/elder/logic/commands/PairCommand.java:122]: Would it be better to commit the Datastore after the personlist is reset? Then when someone undos they get the same view that they see when they previously ran the command.

**Review by chaaaaun — APPROVED** [review]
Overall good clean implementation!

File: src/main/java/scrolls/elder/logic/commands/RedoCommand.java [hunk 4]
```diff
@@ -0,0 +1,25 @@
...
+ */
+public class RedoCommand extends Command {
+    public static final String COMMAND_WORD = "redo";
+    public static final String MESSAGE_SUCCESS = "Previous undo operation has been reversed!";
+    public static final String MESSAGE_REDO_ERROR = "No previous undo operation to be reversed";
+
+    @Override
+    public CommandResult execute(Model model) throws CommandException {
+        requireNonNull(model);
+        if (!model.getDatastoreVersionStorage().canRedo()) {
```
Comment (soons1) [inline src/main/java/scrolls/elder/logic/commands/RedoCommand.java]: Checking if there is a previous operation to be redone is important, but how might that tie in with the assert statement in redoChanges() in terms of 2 layers of checking?

File: src/main/java/scrolls/elder/logic/commands/RedoCommand.java [hunk 5]
```diff
@@ -0,0 +1,25 @@
...
+            throw new CommandException(MESSAGE_REDO_ERROR);
+        }
+        model.redoChanges();
+        return new CommandResult(MESSAGE_SUCCESS);
```
Comment (soons1) [inline src/main/java/scrolls/elder/logic/commands/RedoCommand.java:25]: Should we consider updating the displayed lists after redoing the changes?

File: src/main/java/scrolls/elder/logic/commands/UndoCommand.java [hunk 6]
```diff
@@ -0,0 +1,26 @@
...
+
+    @Override
+    public CommandResult execute(Model model) throws CommandException {
+        requireNonNull(model);
+        if (!model.getDatastoreVersionStorage().canUndo()) {
+            throw new CommandException(MESSAGE_UNDO_ERROR);
+        }
+
+        model.undoChanges();
+        return new CommandResult(MESSAGE_SUCCESS);
```
Comment (soons1) [inline src/main/java/scrolls/elder/logic/commands/UndoCommand.java:25]: Same as RedoCommand, should we update the displayed list after undoing the operation?

File: src/main/java/scrolls/elder/model/DatastoreVersionStorage.java [hunk 7]
```diff
@@ -0,0 +1,109 @@
...
+        return prevDatastore;
+    }
+
+    /**
+     * Executes the redo operation by moving forward to the next version of the Datastore.
+     *
+     * @return The Datastore in its next state after reversing an undo operation.
+     */
+    public ReadOnlyDatastore executeRedo() {
+        assert this.canRedo() : "Redo command cannot be executed";
```
Comment (soons1) [inline src/main/java/scrolls/elder/model/DatastoreVersionStorage.java:83]: If we were to fall to this layer of assertion, is it possible to display the message string to the user on the GUI?

File: src/main/java/scrolls/elder/logic/parser/AddressBookParser.java [hunk 8]
```diff
@@ -92,6 +94,12 @@ public Command parseCommand(String userInput) throws ParseException {
         case LogAddCommand.COMMAND_WORD:
             return new LogAddCommandParser().parse(arguments);
 
+        case UndoCommand.COMMAND_WORD:
```
Comment (soons1) [inline src/main/java/scrolls/elder/logic/parser/AddressBookParser.java:97]: Do we want to consider alternative command words for the UndoCommand? We can possibly consider "reverse", "cancel" etc.

**Review by soons1 — COMMENTED** [review]
LGTM 👍 
Just need to add some test cases to validate some functions in ModelManager, RedoCommand, UndoCommand etc. and it is good to go!

File: src/main/java/scrolls/elder/model/DatastoreVersionStorage.java (hunk 1 above)
Comment (gerteck) [reply src/main/java/scrolls/elder/model/DatastoreVersionStorage.java:70]: Agreed. If I understand this correctly, the logic here will still hold, since `UndoCommand::Execute` has a guard clause checking `canUndo`, which throws an exception if it fails.  Hence, the assertion here is a supplementary sanity check to ensure correct usage of the `executeUndo` method 👍

**Review by gerteck — APPROVED** [review]
LGTM, would be good to add supplementary testcases and it will be ready to merge! 🍔

File: src/main/java/scrolls/elder/model/DatastoreVersionStorage.java (hunk 2 above)
Comment (CJerrong) [reply src/main/java/scrolls/elder/model/DatastoreVersionStorage.java:100]: Yep I will be implementing test cases on undo and redo commands and pushing them to this pr.

File: src/main/java/scrolls/elder/logic/commands/RedoCommand.java (hunk 4 above)
Comment (CJerrong) [reply src/main/java/scrolls/elder/logic/commands/RedoCommand.java]: I treated the assert statement in redoChanges more like a sanity check to ensure correctness, although it might be fine to remove the assertion as well

File: src/main/java/scrolls/elder/logic/commands/UndoCommand.java (hunk 6 above)
Comment (CJerrong) [reply src/main/java/scrolls/elder/logic/commands/UndoCommand.java:25]: That's a good point, maybe it would be be clearer for the user as well, I'll add that in.

File: src/main/java/scrolls/elder/logic/parser/AddressBookParser.java (hunk 8 above)
Comment (CJerrong) [reply src/main/java/scrolls/elder/logic/parser/AddressBookParser.java:97]: That's a possible alternative as well, maybe we can discuss it and I can make the changes accordingly.

File: src/test/java/scrolls/elder/logic/commands/UndoCommandTest.java [hunk 9]
```diff
@@ -0,0 +1,158 @@
...
+import scrolls.elder.testutil.TypicalDatastore;
+import scrolls.elder.testutil.TypicalIndexes;
+import scrolls.elder.testutil.TypicalPersons;
+public class UndoCommandTest {
+    private Model model;
+    private Model expectedModel;
+    private PersonStore personStore;
+    private PersonStore expectedPersonStore;
+
+    @BeforeEach
```
Comment (gerteck) [inline src/test/java/scrolls/elder/logic/commands/UndoCommandTest.java:27]: Nice Usage of `@BeforeEach` here

**Review by gerteck — APPROVED** [review]
Tests LGTM 👍
